src/submission/NodeSubmissionClient.ts

```diff
@@ -16,44 +16,35 @@ export class NodeSubmissionClient extends DefaultSubmissionClient {
     this.configurationVersionHeader = this.configurationVersionHeader.toLowerCase();
   }
 
-  public sendRequest(config:Configuration, method:string, path:string, data:string, callback: (status:number, message:string, data?:string, headers?:Object) => void): void {
-    function complete(response:http.IncomingMessage, responseBody:string, responseHeaders:Object) {
-      var message:string;
-      if (response.statusCode === 0) {
-        message = 'Unable to connect to server.';
-      } else if (response.statusCode < 200 || response.statusCode > 299) {
-        message = response.statusMessage || (<any>response).message;
-      }
+  public sendRequest(config: Configuration, method: string, path: string, data: string, callback: NodeSubmissionCallback): void {
+    var request: NodeSubmissionRequest = {
+      method,
+      path,
+      data,
+      serverUrl: config.serverUrl,
+      apiKey: config.apiKey,
+      userAgent: config.userAgent
+    };
 
-      callback(response.statusCode || 500, message, responseBody, responseHeaders);
+    var exitController = config.exitController;
+    if (exitController.isApplicationExiting) {
+      this.sendRequestSync(request, callback);
+    } else {
+      submitRequest(request, callback);
     }
+  }
 
-    var parsedHost = url.parse(config.serverUrl);
-    var options:https.RequestOptions = {
-      auth: `client:${config.apiKey}`,
-      headers: {},
-      hostname: parsedHost.hostname,
-      method: method,
-      port: parsedHost.port && parseInt(parsedHost.port),
-      path: path
-    };
+  private sendRequestSync(request: NodeSubmissionRequest, callback: NodeSubmissionCallback): void {
```

Comment:
In Node.js, there is only a single thread. As long as there are any open pending callbacks (either timers or I/O), the application will continue. As soon as all callbacks in the queue are processed, the application exits.

There are exceptions to this rule: In case of an uncaught exception or a call to `process.exit()`, no other handlers are processed. Instead, the `process.on("exit")` handlers run, but immediately after the application will exit.

The http API in Node.js is completely asynchronous, meaning that requests are sent using a callback to wait for the response. As we can't wait for callbacks inside `process.on("exit")` we have to use a child process, as `spawnSync` actually allows us to wait for the child process to finish.